pkg/apis/ocs/v1alpha1/storagecluster_types.go

```diff
@@ -10,8 +12,20 @@ import (
 // StorageClusterSpec defines the desired state of StorageCluster
 type StorageClusterSpec struct {
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
-	ManageNodes  bool   `json:"manageNodes,omitempty"`
-	InstanceType string `json:"instanceType,omitempty"`
+	ManageNodes       bool               `json:"manageNodes,omitempty"`
+	InstanceType      string             `json:"instanceType,omitempty"`
+	StorageDeviceSets []StorageDeviceSet `json:"storageDeviceSets"`
+}
+
+// StorageDeviceSet defines a set of storage devices.
```

Comment:
Maybe we should just use `StorageClassDeviceSet` from here, as for now I've just copied over the entire struct. We only really need Count, Resources, Placement and Templates. We could generate Name ourselves and provide any rook-ceph specific Config ourselves.

Reply:
We should not generate the name ourselves. The name needs to be provided so that we can set it as the name of the `StorageClassDeviceSet` in the CephCluster CR. The installer can autogenerate a name for us, but from the operator point of view we should always mandate a name.

Also, we should not be using the `StorageClassDeviceSet` struct directly, since that is a generic Rook struct and we are specifically targeting Rook-Ceph. I would say:
* Replace `VolumeClaimTemplates []corev1.PersistentVolumeClaim` with `DataClaimTemplate corev1.PersistentVolumeClaim`, since we will only support one PVC per pod for our initial implementation. We can define a fixed naming template for this as well, so no need to specify a name for the claim template.
* `Config` should not be a `map[string]string`. Instead it should be its own struct where we explicitly specify the OSD configurations we want to expose.
* We should have a `Capacity` field so the admin doesn't have to specify it under `Resources`.

I've also been wanting to not have the admin specify a full `Placement` spec, opting instead for something like a `PlacementStrategy` that would be a defined set of strings that correspond to well-known and managed `Placement` specs. Unfortunately, I haven't been able to come up with a good scheme for these on my own, so I'd like to discuss it further offline.

Reply:
I wasn't really clear with my comment earlier, but I agree with you about not importing `StorageClassDeviceSet`. 

I prefer to the have `VolumeClaimTemplates` be a slice. We can have an explicit length check, for now, to ensure that it just has 1 template. If we would want to support having fast journal devices for OSDs later on, this will help. I'll change the name as suggested, to make it clear that the templates are for data PVCs only.

For `Config` I'll make it an empty struct for now. We can add in members later. I'll leave `Placement` as is, until we upon the scheme for `PlacementStrategy`.

Reply:
> I prefer to the have `VolumeClaimTemplates` be a slice. We can have an explicit length check, for now, to ensure that it just has 1 template. If we would want to support having fast journal devices for OSDs later on, this will help. I'll change the name as suggested, to make it clear that the templates are for data PVCs only.

Right, my thought for doing that would be we'd add a new field called something like `MetadataClaimTemplate` or `JournalClaimTemplate` to specify the second OSD. While we want to be flexible and extensible in upstream Rook, here we want to be as restrictive and prescriptive as possible. My main concern with the slice is that I don't like relying on a pure numerical index to identify a given device set's purpose, I'd rather be explicit about it's functionality.

> For `Config` I'll make it an empty struct for now. We can add in members later. I'll leave `Placement` as is, until we upon the scheme for `PlacementStrategy`.

Works for me.

Reply:
Explicit fields sounds good to me. I'll make the changes.